### PR TITLE
Generate Inventory & Sha7 appended filenames for binaries

### PR DESCRIPTION
**Pull request by schneems**
The Ruby buildpack is tightly coupled to the URLs generated by this repo. This change introduces an inventory/manifest file that can be used as a layer of indirection. This is a pattern used by several other buildpacks:

- [go](https://github.com/heroku/buildpacks-go/blob/a94957b68151916eba8be74d489389ee0f011592/buildpacks/go/inventory.toml#L1-L6)
- [jvm](https://github.com/heroku/buildpacks-jvm/blob/dd49fc24ef7b740646da91a665a2747fb5553de7/buildpacks/jvm/openjdk_inventory.toml)
- [dotnet](https://github.com/heroku/buildpacks-dotnet/blob/3ac8c9a7460e242d8156b17808b19cdebdfdfa8d/buildpacks/dotnet/inventory.toml)

The file acts as a lookup, so I can say "find the entry with version 3.3.3 from `ruby_inventory.toml` for `heroku-24` with `arm64` CPU architecture and it will return an entry containing a URL and a checksum. The URL says where to find that artifact and the checksum validates that the artifact hasn't been modified between when it was added to the inventory and now.

It also allows us to decouple modifying the S3 bucket with rolling out the version on Heroku via the buildpack. In effect this allows us to make an automated release pipeline (in the future) where source changes can be detected automatically on the ruby-lang source site and built and uploaded to S3. This will trigger a manifest entry to be added to the files which will (eventually) trigger a PR to the Ruby buildpack. A single engineer can review that, approve and merge it, and then deploy the change. Versus today, the process requires two engineers for compliance. (non-automated PRs require approval of a second engineer that didn't write the change).

## Rebuild support

As we roll this out we have to preserve the current binary filenames or it will break existing buildpacks. However we also need to introduce a new filename to allow for multiple binary builds of the same version. Details:

Currently the same version of Ruby can be re-built which will over-write the existing version. This should produce the same output, but we've not done any work on byte-for-byte reproducible builds and even a single bit changed (due to a timestamp, for example) would cause a different SHA. That means that we cannot ship a manifest file that points directly at the current URLs i.e. `https://heroku-buildpack-ruby.s3.us-east-1.amazonaws.com/heroku-24/arm64/ruby-3.3.4.tgz` otherwise if the build is re-run the SHA will change and the inventory will no longer be valid.

There is a need to be able to generate a binary multiple times in the event of a problem. A recent example is https://github.com/heroku/docker-heroku-ruby-builder/pull/49 where a jruby tar file was generated with truncated filenames.

This is solved by appending the first 7 characters of the sha to the file so now when 3.3.4 is built it will generate both:

- `heroku-24/arm64/ruby-3.3.4.tgz`
- `heroku-24/arm64/ruby-3.3.4-<sha7>.tgz` like `heroku-24/arm64/ruby-3.3.4-7bebeee.tgz`

The second tar file's checksum should never change. This will allow us to roll out a manifest file with checksum validation enabled and also allow us to re-generate binaries if needed.

### CODE REVIEW

File: jruby_executable/src/bin/jruby_build.rs [hunk 1]
```diff
@@ -99,8 +114,61 @@ fn jruby_build(args: &Args) -> Result<(), Box<dyn Error>> {
...
+        let tar_path = tar_file.path();
+        let sha = sha256_from_path(tar_path)?;
+        let sha_seven = sha.chars().take(7).collect::<String>();
+        let sha_seven_path = append_filename_with(tar_path, &format!("-{sha_seven}"), ".tgz")?;
+
+        bullet = bullet.sub_bullet(format!("Write {}", sha_seven_path.display(),));
+        fs_err::copy(tar_file.path(), &sha_seven_path)?;
+
+        let timestamp = chrono::Utc::now();
+        for cpu_arch in &[CpuArch::new("amd64")?, CpuArch::new("arm64")?] {
```
Comment (runesoerensen) [inline jruby_executable/src/bin/jruby_build.rs:126]: Have you considered using `inventory::artifact::Arch` here and [here](https://github.com/heroku/docker-heroku-ruby-builder/pull/51/files#diff-1e3d9cad7329371a0f1f11fe6319513f49ce9f9cbe5f8b4d49cbe96752fe01d9R168) instead?

I think that should work "out of the box", while avoiding initializing CpuArch with strings in multiple places, as well as (eliminate the need for [this type conversion implementation](https://github.com/heroku/docker-heroku-ruby-builder/pull/51/files#diff-293260e5637d839b13a4f4a3d15a973ee4e52691440bf6db1dd1855e9bf59ec2R100-R106)).

I think it'd also be worth considering removing `CpuArch` altogether, but that might be a larger refactoring - not sure if it'd be better to do that in a later PR, but at least for this PR it'd probable make more sense to use the types available in `inventory`?

Comment (runesoerensen) [reply jruby_executable/src/bin/jruby_build.rs:126]: I mean something like this :)

```suggestion
        for cpu_arch in [Arch::Amd64, Arch::Arm64] {
```

Comment (runesoerensen) [reply jruby_executable/src/bin/jruby_build.rs:126]: I think it probably would be preferable to remove CpuArch with this change. I pushed a branch with a commit doing that, feel free to cherry-pick that or push that branch to your own. Or delete the branch from this repo if you want

https://github.com/heroku/docker-heroku-ruby-builder/commit/08efc6f4ae7d254f64a099d2200be133f21522ed

Comment (schneems) [reply jruby_executable/src/bin/jruby_build.rs]: Awesome suggestion. I pulled your commit into this branch.

Comment (schneems) [reply jruby_executable/src/bin/jruby_build.rs]: I think that resolves this conversation, closing. CI is now green.

**Review by runesoerensen — APPROVED** [review]
I’d probably have preferred less duplication between `ruby_build.rs` and `jruby_build.rs`. Seems to me those could actually be merged almost entirely, and at least the newly [introduced logic](https://github.com/heroku/docker-heroku-ruby-builder/pull/51/files#diff-a868472eba862d55e56267e2e67aa3cb2c91fea7684c88ec53fe2ac5bf4d2062R156-R208) is identical.

Probably more of a matter of style (and not super important in this context/for what this repository does), just seems like there’s higher than necessary risk that the two implementations will fall out of sync over time.

I’d also have preferred to see all the code being included in this PR (and not in the `GemVersion` crate), but I understand the intention is to move that in the future/this is acceptable for now, so going to approve :)

**Pull request merged**